temporalio/worker/replayer.py

```diff
@@ -0,0 +1,272 @@
+"""Replayer."""
+
+from __future__ import annotations
+
+import concurrent.futures
+import copy
+import json
+import logging
+import re
+from typing import Any, Dict, Iterable, Optional, Type, Union
+
+import google.protobuf.json_format
+from typing_extensions import TypedDict
+
+import temporalio.api.history.v1
+import temporalio.bridge.worker
+import temporalio.converter
+
+from .interceptor import Interceptor
+from .worker import load_default_build_id
+from .workflow import _WorkflowWorker
+from .workflow_instance import UnsandboxedWorkflowRunner, WorkflowRunner
+
+logger = logging.getLogger(__name__)
+
+
+class Replayer:
+    """Replayer to replay workflows from history."""
+
+    def __init__(
+        self,
+        *,
+        workflows: Iterable[Type],
+        workflow_task_executor: Optional[concurrent.futures.ThreadPoolExecutor] = None,
+        workflow_runner: WorkflowRunner = UnsandboxedWorkflowRunner(),
+        namespace: str = "ReplayNamespace",
+        data_converter: temporalio.converter.DataConverter = temporalio.converter.default(),
+        interceptors: Iterable[Interceptor] = [],
+        build_id: Optional[str] = None,
+        identity: Optional[str] = None,
+        debug_mode: bool = False,
+    ) -> None:
+        """Create a replayer to replay workflows from history.
+
+        See :py:meth:`temporalio.worker.Worker.__init__` for a description of
+        arguments.
+        """
+        if not workflows:
+            raise ValueError("At least one workflow must be specified")
+        self._config = ReplayerConfig(
+            workflows=workflows,
+            workflow_task_executor=workflow_task_executor,
+            workflow_runner=workflow_runner,
+            namespace=namespace,
+            data_converter=data_converter,
+            interceptors=interceptors,
+            build_id=build_id,
+            identity=identity,
+            debug_mode=debug_mode,
+        )
+
+    def config(self) -> ReplayerConfig:
+        """Config, as a dictionary, used to create this replayer.
+
+        Returns:
+            Configuration, shallow-copied.
+        """
+        config = self._config.copy()
+        config["workflows"] = list(config["workflows"])
+        return config
+
+    async def replay_workflow(
+        self,
+        history: Union[temporalio.api.history.v1.History, str, Dict[str, Any]],
+    ) -> None:
+        """Replay a workflow for the given history.
+
+        Args:
+            history: The history to replay. Can be a proto history object, a
+                JSON string of the history as exported via web/tctl, or a
+                dictionary of the JSON history.
+        """
```

Comment:
Not clear to me at all what the dict representation is

Reply:
JSON as a dict. Basically what https://docs.python.org/3/library/json.html#json.load would return. I have this in there just in case people wanted to pass around the generic Python structure instead of string or proto. I can clarify.

Reply:
+1 to clarifying

Reply:
Clarified.